### PR TITLE
libp2phttp: HTTP Peer ID Authentication

### PR DESCRIPTION
**Pull request by MarcoPolo**
This enables HTTP peers to authenticate each other's peer ID. This would allow users to use an http transport that has a peer id component (e.g. `/dns/example.com/http/p2p/12Foo`). I think it's nice to have this for completeness so that an http transport has the same semantics as a libp2p stream transport when doing HTTP with regard to Peer IDs.

There's more testing I want to do here, but I think this is more or less ready for a review.

For a high level overview of the authentication protocol refer to the overview in the spec: https://github.com/libp2p/specs/blob/45006f17d2fa0cede50b2db2311a55061011a3fc/http/peer-id-auth.md#mutual-client-and-server-peer-id-authentication-overview

### CODE REVIEW

**Comment by MarcoPolo** [comment]
I can't use base64's appendencode since it was only introduced in go 1.22. Easy enough to backport though.

File: p2p/http/auth/server.go [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+func (a *ServerPeerIDAuth) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	hostname := r.Host
+	if !a.InsecureNoTLS {
+		if r.TLS == nil {
+			log.Debugf("No TLS connection, and InsecureNoTLS is false")
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+		if hostname != r.TLS.ServerName {
+			log.Debugf("Unauthorized request for host %s: hostname mismatch. Expected %s", hostname, r.TLS.ServerName)
```
[the remark this answers is not in the document]
Comment (MarcoPolo) [reply p2p/http/auth/server.go:47]: Do you mean that we should just get rid of the "InsecureNoTLS" option? I'm not sure we should. Users may want to have this running on a setup where something else terminates tls (e.g. nginx proxy)

File: p2p/http/auth/server.go [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+		defer pool.Put(blobB64)
+
+		blobB64 = blobB64[:0]
+		blobB64 = append(blobB64, BearerAuthScheme...)
+		blobB64 = append(blobB64, ' ')
+		blobB64 = b64AppendEncode(blobB64, b)
+
+		w.Header().Set("Authorization", string(blobB64))
+
+		if base64.URLEncoding.DecodedLen(len(f.challengeServerB64)) >= challengeLen {
```
[the remark this answers is not in the document]
Comment (MarcoPolo) [reply p2p/http/auth/server.go:100]: The server doesn't sign anything. Effectively client only auth. I guess it's fine to explicitly error though.

File: p2p/http/auth/server.go [hunk 3]
```diff
@@ -0,0 +1,434 @@
...
+	sig, err := privKey.Sign(buf)
+	if err != nil {
+		return nil, fmt.Errorf("failed to sign bearer token: %w", err)
+	}
+	buf = append(buf, sig...)
+
+	return buf, nil
+}
+
+func parseBearerTokenBlob(privKey crypto.PrivKey, blob []byte) (bearerToken, error) {
```
[the remark this answers is not in the document]
Comment (MarcoPolo) [reply p2p/http/auth/server.go:259]: I like this idea. But I might have to flip it to put the sig in the front.

File: p2p/http/auth/client.go [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+		a.tokenMapMu.Unlock()
+	}
+
+	return serverID, nil
+}
+
+func (a *ClientPeerIDAuth) sign(challengeClientB64 string, hostname string) ([]byte, error) {
+	return sign(a.PrivKey, PeerIDAuthScheme, []string{
+		"challenge-client=" + challengeClientB64,
+		fmt.Sprintf(`hostname="%s"`, hostname),
```
[the remark this answers is not in the document]
Comment (MarcoPolo) [reply p2p/http/auth/client.go:110]: replied in the thread. I'll change it to also quote the challenges too.

File: p2p/http/auth/server.go [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+
+	// Hostname
+	buf = binary.AppendUvarint(buf, uint64(len(t.hostname)))
+	buf = append(buf, t.hostname...)
+
+	// Created at
+	buf = binary.AppendUvarint(buf, uint64(len(createdAtBytes)))
+	buf = append(buf, createdAtBytes...)
+
+	sig, err := privKey.Sign(buf)
```
[the remark this answers is not in the document]
Comment (MarcoPolo) [reply p2p/http/auth/server.go:250]: I'm limiting myself to what is available in `PrivKey` interface (which is just `.Sign`)

[the remark this answers is not in the document]
Comment (MarcoPolo) [reply p2p/http/auth/server.go:250]: ah, I'm seeing your other comment about a new random key. That seems better to me.

File: p2p/http/auth/server.go [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+	partsToVerify = append(partsToVerify, fmt.Sprintf(`hostname="%s"`, f.hostname))
+
+	err = verifySig(f.pubKey, PeerIDAuthScheme, partsToVerify, f.signature)
+	if err != nil {
+		return "", fmt.Errorf("failed to verify signature: %s", err)
+	}
+	return peer.IDFromPublicKey(f.pubKey)
+}
+
+func (a *ServerPeerIDAuth) UnwrapBearerToken(r *http.Request, expectedHostname string) (peer.ID, error) {
```
[the remark this answers is not in the document]
Comment (MarcoPolo) [reply p2p/http/auth/server.go]: I switched to JSON. It introduces some allocs I don't love, but it's definitely easier.

File: p2p/http/auth/server.go [hunk 7]
```diff
@@ -0,0 +1,434 @@
...
+	if !a.InsecureNoTLS {
+		if r.TLS == nil {
+			log.Debugf("No TLS connection, and InsecureNoTLS is false")
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+		if hostname != r.TLS.ServerName {
+			log.Debugf("Unauthorized request for host %s: hostname mismatch. Expected %s", hostname, r.TLS.ServerName)
+			w.WriteHeader(http.StatusBadRequest)
+			return
```
[the remark this answers is not in the document]
Comment (MarcoPolo) [reply p2p/http/auth/server.go]: Is there a way to terminate the connection that works across http versions? h2 connections don't seem to support http.Hijacker

**Comment by MarcoPolo** [comment]
I've completely refactored this. The handshake logic is now neatly in `internal/handshake`. The Client API is simpler, and the Server API changed a little bit (in a way that makes it hopefully easier to use). The server uses HMAC to authenticate the opaque and the token. And it's also close to 10x faster (still could be improved, but this might be good enough for now).

The main missing thing is the test that generates the examples for the spec. I'll work on that next, but otherwise I think this is ready.

when you get a chance, could I get a review here and in the spec https://github.com/libp2p/specs/pull/564 @sukunrt .

**Review by sukunrt — APPROVED** [review]

**Comment by sukunrt** [comment]
The new spec is much nicer! Thanks @MarcoPolo

File: p2p/http/auth/client.go [hunk 8]
```diff
@@ -1,58 +1,101 @@
...
+		handshake.ParseHeader(resp.Header)
 	}
+	originalBody := req.Body
 
+	handshake.Run()
+	handshake.SetHeader(req.Header)
+
+	// Don't send the body before we've authenticated the server
+	req.Body = nil
 	resp, err := client.Do(req)
```
Comment (aschmahmann) [inline p2p/http/auth/client.go:92]: I don't think this is quite enough:
- ContentLength is already set which means we'll get an error here (adding a test with a non-nil body should show this)
- If GetBody is set would the Body end up getting sent before authorization if we get a 307/308?

Comment (MarcoPolo) [reply p2p/http/auth/client.go:92]: Thanks

File: p2p/http/auth/client.go [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+
+	// Don't send the body before we've authenticated the server
+	req.Body = nil
+	resp, err := client.Do(req)
+	if err != nil {
+		return "", nil, err
+	}
+	resp.Body.Close()
+
+	err = handshake.ParseHeader(resp.Header)
```
[the remark this answers is not in the document]
Comment (MarcoPolo) [reply p2p/http/auth/client.go]: On the client? Or do you mean assert the status code is 401?

File: p2p/http/auth/client.go [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+	// We need to make one final request with the body now that we authenticated
+	// the server.
+	req = req.Clone(req.Context())
+	req.Body = originalBody
+	handshake.SetHeader(req.Header)
+	resp, err = client.Do(req)
+	if err != nil {
+		return "", nil, fmt.Errorf("failed to do authenticated request: %w", err)
+	}
+	return serverPeerID, resp, nil
```
[the remark this answers is not in the document]
Comment (MarcoPolo) [reply p2p/http/auth/client.go]: I've changed this to be a for loop instead since it's all the same thing.
